pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,628 @@
+				if amount > budget {
+					// Run out of budget, let's check if the request is still in the grace period
+					return now - request.start_time > grace_period;
+				} else {
+					// Otherwise we allocate some budget to virtually fulfill the request
+					budget -= amount;
+				}
+			}
+			false
+		}
+
+		/// Mint a new nft in the Pool's collection and store some shares in it
+		pub fn add_stake_to_new_nft(
+			pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			account_id: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> BalanceOf<T> {
+			let shares = match pool.share_price() {
+				Some(price) if price != fp!(0) => bdiv(amount, &price),
+				_ => amount, // adding new stake (share price = 1)
+			};
+			Self::mint_nft(pool.cid, account_id, shares, amount)
+				.expect("mint should always success; qed.");
+			pool.total_shares += shares;
+			pool.total_value += amount;
+			pool.free_stake += amount;
+			shares
+		}
+
+		/// Remove some stakes from nft when withdraw or process_withdraw_queue called.
+		pub fn remove_stake_from_nft(
+			pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			shares: BalanceOf<T>,
+			nft: &mut NftAttr<BalanceOf<T>>,
+		) -> Option<(BalanceOf<T>, BalanceOf<T>, BalanceOf<T>)> {
+			let price = pool.share_price()?;
+			let amount = bmul(shares, &price);
+
+			let amount = amount.min(pool.free_stake);
+
+			let user_shares = nft.shares.checked_sub(&shares)?;
+			let (user_shares, shares_dust) = extract_dust(user_shares);
+
+			let removed_shares = shares + shares_dust;
+			let total_shares = pool.total_shares.checked_sub(&removed_shares)?;
+
+			let (total_stake, _) = extract_dust(pool.total_value - amount);
+			if total_stake > Zero::zero() {
+				pool.free_stake -= amount;
+				pool.total_value -= amount;
+			} else {
+				pool.free_stake = Zero::zero();
+				pool.total_value = Zero::zero();
+			}
+			pool.total_shares = total_shares;
+			let user_dust = nft.remove_stake(user_shares);
+			nft.shares = user_shares;
+
+			Some((amount, user_dust, removed_shares))
+		}
+
+		#[frame_support::transactional]
+		pub fn mint_nft(
+			cid: CollectionId,
+			contributer: T::AccountId,
+			shares: BalanceOf<T>,
+			stakes: BalanceOf<T>,
+		) -> Result<NftId, DispatchError> {
+			pallet_rmrk_core::Collections::<T>::get(cid)
+				.ok_or(pallet_rmrk_core::Error::<T>::CollectionUnknown)?;
+			let nft_id = pallet_rmrk_core::NextNftId::<T>::get(cid);
+
+			pallet_rmrk_core::Pallet::<T>::mint_nft(
+				Origin::<T>::Signed(pallet_id()).into(),
+				Some(contributer),
+				cid,
+				None,
+				None,
+				Default::default(),
+				true,
+				None,
+			)?;
+
+			let attr = NftAttr { shares, stakes };
+			//TODO(mingxuan): an external lock is needed to protect nft_attr from dirty write.
+			Self::set_nft_attr(cid, nft_id, &attr)?;
+			pallet_rmrk_core::Pallet::<T>::set_lock((cid, nft_id), true);
+			Ok(nft_id)
+		}
+
+		#[frame_support::transactional]
+		pub fn burn_nft(cid: CollectionId, nft_id: NftId) -> DispatchResult {
```

Comment:
It depends on what condition burn_nft called, if we call it just for merge, not shares should be removed, when processing withdraw queue, function remove_stake_from_nft will deal with pool shares, and if the shares in nft is zero or dust, it will be burnt.